File: adapters/curl-impersonate/internal/logging/logging_test.go
// SPDX-License-Identifier: Apache-2.0

package logging

import (
	"bytes"
	"context"
	"encoding/json"
	"testing"

	"go.opentelemetry.io/otel/trace"
)

func TestNew_EmitsElevenMandatoryFieldsAtInfoLevel(t *testing.T) {
	var buf bytes.Buffer
	logger := New(&buf, "spectre-curl-impersonate", "test-version")

	logger.Info("hello", "job_id", "job-1", "request_id", "req-1")

	var obj map[string]any
	if err := json.Unmarshal(buf.Bytes(), &obj); err != nil {
		t.Fatalf("expected JSON output, got %q: %v", buf.String(), err)
	}

	// ADR-0031 §3.4 mandates 11 fields. tenant_id is always null
	// in v1alpha1; trace_id / span_id are null when no OTel
	// context is active (this case).
	for _, k := range []string{
		"timestamp",
		"level",
		"service",
		"service_version",
		"caller",
		"message",
		"trace_id",
		"span_id",
		"request_id",
		"job_id",
		"tenant_id",
	} {
		if _, ok := obj[k]; !ok {
			t.Errorf("missing field %q in %v", k, obj)
		}
	}
	if obj["level"] != "INFO" {
		t.Errorf("level = %v; want INFO", obj["level"])
	}
	if obj["service"] != "spectre-curl-impersonate" {
		t.Errorf("service = %v; want spectre-curl-impersonate", obj["service"])
	}
	if obj["service_version"] != "test-version" {
		t.Errorf("service_version = %v; want test-version", obj["service_version"])
	}
	if obj["message"] != "hello" {
		t.Errorf("message = %v; want hello", obj["message"])
	}
	if obj["job_id"] != "job-1" {
		t.Errorf("job_id = %v; want job-1", obj["job_id"])
	}
	if obj["request_id"] != "req-1" {
		t.Errorf("request_id = %v; want req-1", obj["request_id"])
	}
	if obj["tenant_id"] != nil {
		t.Errorf("tenant_id = %v; want null", obj["tenant_id"])
	}
	if obj["trace_id"] != nil {
		t.Errorf("trace_id = %v; want null (no active span)", obj["trace_id"])
	}
	if obj["span_id"] != nil {
		t.Errorf("span_id = %v; want null (no active span)", obj["span_id"])
	}
}

func TestNew_PopulatesTraceIdFromActiveSpanContext(t *testing.T) {
	var buf bytes.Buffer
	logger := New(&buf, "spectre-curl-impersonate", "test-version")

	traceID, _ := trace.TraceIDFromHex("0123456789abcdef0123456789abcdef")
	spanID, _ := trace.SpanIDFromHex("0123456789abcdef")
	sc := trace.NewSpanContext(trace.SpanContextConfig{
		TraceID:    traceID,
		SpanID:     spanID,
		TraceFlags: trace.FlagsSampled,
		Remote:     true,
	})
	ctx := trace.ContextWithSpanContext(context.Background(), sc)

	logger.InfoContext(ctx, "inside span")

	var obj map[string]any
	if err := json.Unmarshal(buf.Bytes(), &obj); err != nil {
		t.Fatalf("expected JSON: %v", err)
	}
	if got := obj["trace_id"]; got != "0123456789abcdef0123456789abcdef" {
		t.Errorf("trace_id = %v; want 0123456789abcdef0123456789abcdef", got)
	}
	if got := obj["span_id"]; got != "0123456789abcdef" {
		t.Errorf("span_id = %v; want 0123456789abcdef", got)
	}
}

func TestFormatSource_StripsDirectoryPrefix(t *testing.T) {
	// Black-box: invoke through the logger so the slog.Record
	// gets a real Source, then assert the rendered caller is the
	// basename:line form ADR-0031 §3.4 prescribes.
	var buf bytes.Buffer
	logger := New(&buf, "svc", "v")
	logger.Info("checkpoint") // recorded on this file:<line>

	var obj map[string]any
	if err := json.Unmarshal(buf.Bytes(), &obj); err != nil {
		t.Fatalf("expected JSON: %v", err)
	}
	caller, _ := obj["caller"].(string)
	if caller == "" {
		t.Fatalf("caller missing")
	}
	// Verify no directory separator in the rendered string.
	for _, c := range caller {
		if c == '/' || c == '\\' {
			t.Errorf("caller %q contains a path separator; expected basename only", caller)
		}
	}
}

File: adapters/playwright/src/logging.test.ts
// SPDX-License-Identifier: Apache-2.0
//
// W3.2 Cluster A tests for `./logging`.

import { describe, expect, it, vi } from "vitest";
import { Writable } from "node:stream";
import pino from "pino";
import {
  trace,
  TraceFlags,
  type Span,
  type SpanContext,
} from "@opentelemetry/api";

// Re-implement `createLogger`'s `formatters.log` so we can target
// an in-memory writable buffer (rather than stdout via
// `pino.destination`) and assert on captured bytes. Mirrors the
// `createLogger` implementation 1:1 — if it drifts the
// `expect(obj.service).toBe(...)` assertion fails and the test
// flags the divergence.
function loggerWithBuffer(serviceVersion: string): {
  logger: pino.Logger;
  read: () => string;
} {
  let captured = "";
  const sink = new Writable({
    write(chunk: Buffer, _enc, cb) {
      captured += chunk.toString();
      cb();
    },
  });
  const logger = pino(
    {
      level: "trace",
      base: null,
      timestamp: () => `,"timestamp":"${new Date().toISOString()}"`,
      messageKey: "message",
      formatters: {
        level: (label) => ({ level: label.toUpperCase() }),
        log(object) {
          const span = trace.getActiveSpan();
          const sc = span?.spanContext();
          const traceFields =
            sc && sc.traceId && sc.spanId
              ? { trace_id: sc.traceId, span_id: sc.spanId }
              : { trace_id: null, span_id: null };
          return {
            service: "spectre-playwright",
            service_version: serviceVersion,
            request_id: null,
            job_id: null,
            tenant_id: null,
            ...traceFields,
            ...object,
          };
        },
      },
    },
    sink,
  );
  return { logger, read: () => captured };
}

describe("createLogger", () => {
  it("emits eleven mandatory fields with nulls for unset slots", () => {
    const { logger, read } = loggerWithBuffer("test-version");
    logger.info({ request_id: "req-1", job_id: "job-1" }, "hello");

    const obj = JSON.parse(read().trim()) as Record<string, unknown>;

    for (const key of [
      "timestamp",
      "level",
      "service",
      "service_version",
      "message",
      "trace_id",
      "span_id",
      "request_id",
      "job_id",
      "tenant_id",
    ]) {
      expect(obj).toHaveProperty(key);
    }
    expect(obj.level).toBe("INFO");
    expect(obj.service).toBe("spectre-playwright");
    expect(obj.service_version).toBe("test-version");
    expect(obj.message).toBe("hello");
    expect(obj.request_id).toBe("req-1");
    expect(obj.job_id).toBe("job-1");
    expect(obj.tenant_id).toBeNull();
    expect(obj.trace_id).toBeNull();
    expect(obj.span_id).toBeNull();
  });

  it("populates trace_id + span_id from the active span context", () => {
    // Stub `trace.getActiveSpan()` directly so the test doesn't
    // depend on a registered context manager (`AsyncHooksContextManager`).
    // Cluster A's production wiring registers one via
    // `NodeSDK.start()` + `HttpInstrumentation`; the unit test
    // verifies the formatter's read path, not the SDK's bookkeeping.
    const synthetic: SpanContext = {
      traceId: "0123456789abcdef0123456789abcdef",
      spanId: "0123456789abcdef",
      traceFlags: TraceFlags.SAMPLED,
      isRemote: true,
    };
    const wrapped = trace.wrapSpanContext(synthetic);
    const spy = vi
      .spyOn(trace, "getActiveSpan")
      .mockReturnValue(wrapped as Span);

    try {
      const { logger, read } = loggerWithBuffer("test-version");
      logger.info("inside span");

      const line = read().trim().split("\n").pop();
      expect(line).toBeDefined();
      const obj = JSON.parse(line as string) as Record<string, unknown>;
      expect(obj.trace_id).toBe(synthetic.traceId);
      expect(obj.span_id).toBe(synthetic.spanId);
    } finally {
      spy.mockRestore();
    }
  });
});
